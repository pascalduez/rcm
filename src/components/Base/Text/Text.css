/* -------------------- Font size ------------------------------------------- */

.fontSizeSmall {
  font-size: $font-size-small-mobile;

  @media screen and (min-width: $screen-small) {
    font-size: $font-size-small;
  }
}

.fontSizeBase {
  font-size: $font-size-base-mobile;

  @media screen and (min-width: $screen-small) {
    font-size: $font-size-base;
  }
}

.fontSizeBig {
  font-size: $font-size-big-mobile;

  @media screen and (min-width: $screen-small) {
    font-size: $font-size-big;
  }
}

.fontSizeBigger {
  font-size: $font-size-bigger-mobile;

  @media screen and (min-width: $screen-small) {
    font-size: $font-size-bigger;
  }
}

/* -------------------- Font weight ----------------------------------------- */

.fontWeightNormal {
  font-weight: 300;
}

.fontWeightBold {
  font-weight: 400;
}

/* -------------------- Font style ------------------------------------------ */

.fontStyleNormal {
  font-style: normal;
}

.fontStyleItalic {
  font-style: italic;
}

.fontStyleOblique {
  font-style: oblique;
}

/* -------------------- Color ----------------------------------------------- */

.colorInherit {
  color: inherit;
}

.colorBase {
  color: $color-text;
}

.colorLight {
  color: $color-text-light;
}

.colorWhite {
  color: white;
}

.colorGreyLighter {
  color: $color-grey-lighter;
}

.colorGreyLight {
  color: $color-grey-light;
}

.colorGrey {
  color: $color-grey;
}

.colorGreyDark {
  color: $color-grey-dark;
}

.colorGreyDarker {
  color: $color-grey-darker;
}

.colorInfo {
  color: $color-info;
}

.colorSuccess {
  color: $color-success;
}

.colorWarning {
  color: $color-warning;
}

.colorDanger {
  color: $color-danger;
}

.colorDisabled {
  color: $color-grey-darker;
}

/* -------------------- Color ----------------------------------------------- */

.hoverColorInherit {
  &:hover {
    color: inherit;
  }
}

.hoverColorWhite {
  &:hover {
    color: white;
  }
}

/* -------------------- Text align ------------------------------------------ */

.textAlignStart {
  text-align: left;
  text-align: start;
}

.textAlignEnd {
  text-align: right;
  text-align: end;
}

.textAlignCenter {
  text-align: center;
}

/* -------------------- Text transform -------------------------------------- */

.textTransformCapitallize {
  text-transform: capitalize;
}

.textTransformLowercase {
  text-transform: lowercase;
}

.textTransformUppercase {
  text-transform: uppercase;
}

/* -------------------- Text decoration ------------------------------------- */

.textDecorationOverline {
  text-decoration: overline;
}

.textDecorationUnderline {
  text-decoration: underline;
}

.textDecorationLineThrough {
  text-decoration: line-through;
}

/* -------------------- Text overflow --------------------------------------- */

.truncate {
  max-width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.break {
  max-width: 100%;
  word-break: break-all;
}

/* -------------------- Display --------------------------------------------- */

.displayInline {
  display: inline;
}

.displayInlineBlock {
  display: inline-block;
}

.displayBlock {
  display: block;
}

/* -------------------- User select ----------------------------------------- */

.selectAuto {
  user-select: auto;
}

.selectText {
  user-select: text;
}

.selectNone {
  user-select: none;
}

.selectContain {
  user-select: contain;
}

.selectAll {
  user-select: all;
}
