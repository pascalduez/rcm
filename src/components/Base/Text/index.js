// @flow

export { default as withText } from './withText';
export type {
  Props as withTextProps,
  Color as TextColor,
  Theme as TextTheme,
} from './withText';
export { default as Text } from './Text';
export type { Props as TextProps } from './Text';
