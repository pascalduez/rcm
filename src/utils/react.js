// @flow

import type { Element } from 'react';

export const getElementDisplayName = (element: Element<*>): string =>
  element.type.displayName ||
  // Function name.
  element.type.name ||
  // Function without a name.
  (typeof element.type === 'function' ? 'No Display Name' : element.type);
