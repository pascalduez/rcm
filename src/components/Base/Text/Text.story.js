// @flow

import * as React from 'react';
import Text from './Text';

export default () => ();
