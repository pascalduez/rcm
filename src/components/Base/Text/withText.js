/* @flow */
/* eslint-disable react/no-unused-prop-types */

import * as React from 'react';
import classNames from 'classnames';

import styles from './Text.css';

type Size = 'small' | 'base' | 'big' | 'bigger';

type Weight = 'normal' | 'bold';

type Style = 'normal' | 'italic' | 'oblique';

type Align = 'start' | 'end' | 'center';

type Transform = 'capitalize' | 'lowercase' | 'uppercase';

type Decoration = 'underline' | 'overline' | 'line-through';

type Overflow = 'truncate' | 'break';

type Display = 'inline' | 'inline-block' | 'block';

type Select = 'auto' | 'text' | 'none' | 'contain' | 'all';

export type Theme = 'info' | 'success' | 'warning' | 'danger' | 'disabled';

export type Color =
  | 'inherit'
  | 'base'
  | 'light'
  | 'white'
  | 'grey-lighter'
  | 'grey-light'
  | 'grey'
  | 'grey-dark'
  | 'grey-darker'
  | 'info'
  | 'success'
  | 'warning'
  | 'danger'
  | 'disabled';

export type HoverColor = 'inherit' | 'white' | 'blue';

type ThemeProps = {
  [key: Theme]: {
    color: Color,
    weight: Weight,
  },
};

const themesMap: ThemeProps = {
  info: {
    color: 'info',
    weight: 'bold',
  },
  success: {
    color: 'success',
    weight: 'bold',
  },
  warning: {
    color: 'warning',
    weight: 'bold',
  },
  danger: {
    color: 'danger',
    weight: 'bold',
  },
  disabled: {
    color: 'disabled',
    weight: 'bold',
  },
};

export type TextProps = {
  size?: ?Size,
  weight?: ?Weight,
  fontStyle?: ?Style,
  color?: ?Color,
  hoverColor?: ?HoverColor,
  align?: ?Align,
  transform?: ?Transform,
  decoration?: ?Decoration,
  theme?: ?Theme,
  overflow?: ?Overflow,
  display?: ?Display,
  select?: ?Select,
  className?: string,
};

/**
 * Helper: font and text styling component.
 */
export default function withText<Props: TextProps>(
  Component: React.ComponentType<Props>
): React.ComponentType<Props> {
  function Text(props: Props) {
    const themeProps = props.theme ? themesMap[props.theme] : {};
    const {
      size,
      weight,
      fontStyle,
      color,
      hoverColor,
      align,
      transform,
      decoration,
      overflow,
      display,
      select,
      theme,
      className,
      ...extraProps
    }: Props = { ...themeProps, ...props };

    const rootClasses = classNames(
      {
        [styles.fontSizeSmall]: size === 'small',
        [styles.fontSizeBase]: size === 'base',
        [styles.fontSizeBig]: size === 'big',
        [styles.fontSizeBigger]: size === 'bigger',

        [styles.fontWeightNormal]: weight === 'normal',
        [styles.fontWeightBold]: weight === 'bold',

        [styles.fontStyleNormal]: fontStyle === 'normal',
        [styles.fontStyleItalic]: fontStyle === 'italic',
        [styles.fontStyleOblique]: fontStyle === 'oblique',

        [styles.colorInherit]: color === 'inherit',
        [styles.colorBase]: color === 'base',
        [styles.colorLight]: color === 'light',
        [styles.colorWhite]: color === 'white',
        [styles.colorGreyLighter]: color === 'grey-lighter',
        [styles.colorGreyLight]: color === 'grey-light',
        [styles.colorGrey]: color === 'grey',
        [styles.colorGreyDark]: color === 'grey-dark',
        [styles.colorGreyDarker]: color === 'grey-darker',
        [styles.colorInfo]: color === 'info',
        [styles.colorSuccess]: color === 'success',
        [styles.colorWarning]: color === 'warning',
        [styles.colorDanger]: color === 'danger',
        [styles.colorDisabled]: color === 'disabled',

        [styles.hoverColorInherit]: hoverColor === 'inherit',
        [styles.hoverColorWhite]: hoverColor === 'white',

        [styles.textAlignStart]: align === 'start',
        [styles.textAlignEnd]: align === 'end',
        [styles.textAlignCenter]: align === 'center',

        [styles.textTransformCapitallize]: transform === 'capitalize',
        [styles.textTransformLowercase]: transform === 'lowercase',
        [styles.textTransformUppercase]: transform === 'uppercase',

        [styles.textDecorationOverline]: decoration === 'overline',
        [styles.textDecorationUnderline]: decoration === 'underline',
        [styles.textDecorationLineThrough]: decoration === 'line-through',

        [styles.truncate]: overflow === 'truncate',
        [styles.break]: overflow === 'break',

        [styles.displayInline]: display === 'inline',
        [styles.displayInlineBlock]: display === 'inline-block',
        [styles.displayBlock]: display === 'block',

        [styles.selectAuto]: select === 'auto',
        [styles.selectText]: select === 'text',
        [styles.selectNone]: select === 'none',
        [styles.selectContain]: select === 'contain',
        [styles.selectAll]: select === 'all',
      },
      className
    );

    return <Component className={rootClasses} {...extraProps} />;
  }

  return Text;
}
