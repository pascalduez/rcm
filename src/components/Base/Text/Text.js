// @flow

import * as React from 'react';
import classNames from 'classnames';

import withText, { type TextProps } from './withText';

export type Props = {
  tagName: string,
  component?: ?React.Element<*>,
  className?: string,
  children?: React.Node,
} & TextProps;

function Text({
  tagName = 'span',
  component,
  className,
  children,
  ...extraProps
}: Props) {
  if (component && React.isValidElement(component)) {
    return React.cloneElement(component, {
      className: classNames(component.props.className, className),
      ...extraProps,
    });
  }

  return React.createElement(tagName, { className, ...extraProps }, children);
}

const Typed = withText(Text);

Text.displayName = 'Text';
Typed.displayName = 'Text';

export default Typed;
