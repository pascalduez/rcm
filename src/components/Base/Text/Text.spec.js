// @flow

import * as React from 'react';
import { render } from 'enzyme';

import Text from './Text';

describe('<Text />', () => {
  it('should render the proper default markup and classes', () => {
    const wrapper = render(<Text>Just a text</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the proper markup following the `tagName` property', () => {
    const wrapper = render(<Text tagName="h1">Just a text</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should add the proper classes based on props', () => {
    const wrapper = render(
      <Text
        size="big"
        weight="bold"
        color="green"
        align="end"
        transform="capitalize"
      >
        Just a text
      </Text>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should pass extra props through', () => {
    const wrapper = render(
      <Text style={{ display: 'block' }}>Just a text</Text>
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('the `component` property', () => {
    it('should inject classes to passed elements', () => {
      const wrapper = render(
        <Text size="bigger" weight="bold" component={<div>Just a text</div>} />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should preserve elements own classes', () => {
      const wrapper = render(
        <Text
          size="bigger"
          weight="bold"
          component={<div className="test">Just a text</div>}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
