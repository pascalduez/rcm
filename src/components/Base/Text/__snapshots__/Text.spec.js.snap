// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should add the proper classes based on props 1`] = `
<span
  class="Text-fontSizeSmall Text-fontWeightBold Text-colorWhite Text-textAlignEnd Text-textTransformCapitallize"
>
  Just a text
</span>
`;

exports[`<Text /> should pass extra props through 1`] = `
<span
  class=""
  style="display:block"
>
  Just a text
</span>
`;

exports[`<Text /> should render the proper default markup and classes 1`] = `
<span
  class=""
>
  Just a text
</span>
`;

exports[`<Text /> should render the proper markup following the \`tagName\` property 1`] = `
<h1
  class=""
>
  Just a text
</h1>
`;

exports[`<Text /> the \`component\` property should inject classes to passed elements 1`] = `
<div
  class="Text-fontSizeBigger Text-fontWeightBold"
>
  Just a text
</div>
`;

exports[`<Text /> the \`component\` property should preserve elements own classes 1`] = `
<div
  class="test Text-fontSizeBigger Text-fontWeightBold"
>
  Just a text
</div>
`;
